README.md
# Apple Photos App plugin for Kodi

This plugin shows an Apple Photos App library in Kodi.  After reading the library database, you will see categories that correspond with their application counterparts:

* Moments
* Albums
* Videos
* Slideshows

Your photo library must be reachable over the filesystem from the machine on which you run Kodi.

## Configuration
The plugin needs to know where your Photo Library file is.  The default ("~/Pictures/Photos Library.photoslibrary") is probably not working in all cases. So, please enter the correct path in the plugin's settings dialog.

## License
GPL v2


CHANGELOG
v1.3.0 (2020-04-28)
- Add viewtype setting

v1.2.0 (2020-04-26)
- Port code to Python 3 to support Kodi v19

v1.1.1 (2020-04-26)
- Add option for displaying edited versions of photos/videos
- Add Chinese translation
- Add Slideshows browsing
- Update Kodi APIs. Now only compatible with Kodi v13+

v1.1.0 (2020-04-23)
- Update for Photos.app version 5.0 database.
- Remove Places/People due to difficulty in implementation.

v1.0 (2017-05-01)
リリース


resources/lib/db.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals
from future import standard_library
standard_library.install_aliases()
from builtins import *
from builtins import object
try:
    from sqlite3 import dbapi2 as sqlite
except ImportError:
    from pysqlite2 import dbapi2 as sqlite

import xbmc

class DB(object):

    def __init__(self, dbfile):
        self.OpenDB(dbfile)

    def __del__(self):
        self.CloseDB()

    def OpenDB(self, dbfile):
        try:
            self.dbconn = sqlite.connect(dbfile)
        except Exception:
            pass

    def CloseDB(self):
        try:
            self.dbconn.close()
        except Exception:
            pass

    def GetMomentList(self, year, month):
        moment_list = []
        cur = self.dbconn.cursor()
        try:
            if year is None:
                cur.execute("""SELECT strftime(
                                 '%Y', a.ZDATECREATED + b.ZTIMEZONEOFFSET + 978307200, 'unixepoch'
                                 ) AS y
                               FROM ZGENERICASSET a, ZADDITIONALASSETATTRIBUTES b
                               WHERE b.ZASSET = a.Z_PK
                                 AND a.ZCLOUDDELETESTATE = 0
                                 AND a.ZCOMPLETE = 1
                                 AND a.ZHIDDEN = 0
                                 AND a.ZTRASHEDSTATE = 0
                               GROUP BY y
                               ORDER BY y DESC
                               """)
            elif month is None:
                cur.execute("""SELECT strftime(
                                 '%m', a.ZDATECREATED + b.ZTIMEZONEOFFSET + 978307200, 'unixepoch'
                                 ) AS m
                               FROM ZGENERICASSET a, ZADDITIONALASSETATTRIBUTES b
                               WHERE b.ZASSET = a.Z_PK
                                 AND a.ZCLOUDDELETESTATE = 0
                                 AND a.ZCOMPLETE = 1
                                 AND a.ZHIDDEN = 0
                                 AND a.ZTRASHEDSTATE = 0
                                 AND strftime(
                                   '%Y', a.ZDATECREATED + b.ZTIMEZONEOFFSET + 978307200, 'unixepoch'
                                   ) = ?
                               GROUP BY m
                               ORDER BY m ASC
                               """, ('%s' % (year[0]),))
            else:
                cur.execute("""SELECT strftime(
                                 '%d', a.ZDATECREATED + b.ZTIMEZONEOFFSET + 978307200, 'unixepoch'
                                 ) AS d
                               FROM ZGENERICASSET a, ZADDITIONALASSETATTRIBUTES b
                               WHERE b.ZASSET = a.Z_PK
                                 AND a.ZCLOUDDELETESTATE = 0
                                 AND a.ZCOMPLETE = 1
                                 AND a.ZHIDDEN = 0
                                 AND a.ZTRASHEDSTATE = 0
                                 AND strftime(
                                   '%Y-%m', a.ZDATECREATED + b.ZTIMEZONEOFFSET + 978307200, 'unixepoch'
                                   ) = ?
                               GROUP BY d
                               ORDER BY d ASC
                               """, ('%s-%s' % (year[0], month[0]),))
            for row in cur:
                moment_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.db: GetMomentList: ' + e, xbmc.LOGERROR)
        cur.close()
        return moment_list

    def GetFolderList(self, folderUuid):
        folder_list = []
        cur = self.dbconn.cursor()
        try:
            if folderUuid == 'root':
                cur.execute("""SELECT f.ZTITLE, f.ZUUID
                               FROM ZGENERICALBUM f, ZGENERICALBUM p
                               WHERE f.ZCLOUDDELETESTATE = 0
                                 AND f.ZTRASHEDSTATE = 0
                                 AND f.ZKIND = 4000
                                 AND f.ZPARENTFOLDER = p.Z_PK
                                 AND p.ZKIND = 3999
                               ORDER BY f.ZTITLE ASC
                               """)
            else:
                cur.execute("""SELECT f.ZTITLE, f.ZUUID
                               FROM ZGENERICALBUM f, ZGENERICALBUM p
                               WHERE f.ZCLOUDDELETESTATE = 0
                                 AND f.ZTRASHEDSTATE = 0
                                 AND f.ZKIND = 4000
                                 AND f.ZPARENTFOLDER = p.Z_PK
                                 AND p.ZUUID = ?
                               ORDER BY f.ZTITLE ASC
                               """, (folderUuid,))
            for row in cur:
                folder_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.db: GetFolderList: ' + e, xbmc.LOGERROR)
        cur.close()
        return folder_list

    def GetAlbumList(self, folderUuid):
        album_list = []
        cur = self.dbconn.cursor()
        try:
            if folderUuid == 'root':
                cur.execute("""SELECT a.ZTITLE, a.ZUUID
                               FROM ZGENERICALBUM a, ZGENERICALBUM f
                               WHERE a.ZCLOUDDELETESTATE = 0
                                 AND a.ZTRASHEDSTATE = 0
                                 AND a.ZKIND = 2
                                 AND a.ZPARENTFOLDER = f.Z_PK
                                 AND f.ZKIND = 3999
                               ORDER BY a.ZTITLE ASC
                               """)
            else:
                cur.execute("""SELECT a.ZTITLE, a.ZUUID
                               FROM ZGENERICALBUM a, ZGENERICALBUM f
                               WHERE a.ZCLOUDDELETESTATE = 0
                                 AND a.ZTRASHEDSTATE = 0
                                 AND a.ZKIND = 2
                                 AND a.ZPARENTFOLDER = f.Z_PK
                                 AND f.ZUUID = ?
                               ORDER BY a.ZTITLE ASC
                               """, (folderUuid,))
            for row in cur:
                album_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.db: GetAlbumList: ' + e, xbmc.LOGERROR)
        cur.close()
        return album_list

    def GetSlideshowList(self):
        slideshow_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute("""SELECT a.ZTITLE, a.ZUUID
                           FROM ZGENERICALBUM a, ZGENERICALBUM f
                           WHERE a.ZCLOUDDELETESTATE = 0
                             AND a.ZTRASHEDSTATE = 0
                             AND a.ZKIND = 1508
                             AND a.ZPARENTFOLDER = f.Z_PK
                             AND f.ZKIND = 3998
                           ORDER BY a.ZTITLE ASC
                           """)
            for row in cur:
                slideshow_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.db: GetSlideshowList: ' + e, xbmc.LOGERROR)
        cur.close()
        return slideshow_list

    def GetVideoList(self):
        video_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                  m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                           FROM ZGENERICASSET m, ZADDITIONALASSETATTRIBUTES b
                           WHERE m.ZCLOUDDELETESTATE = 0
                             AND m.ZCOMPLETE = 1
                             AND m.ZHIDDEN = 0
                             AND m.ZTRASHEDSTATE = 0
                             AND m.ZKIND = 1
                             AND b.ZASSET = m.Z_PK
                           ORDER BY m.ZDATECREATED ASC
                           """)
            for row in cur:
                video_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.db: GetVideoList: ' + e, xbmc.LOGERROR)
        cur.close()
        return video_list

    def GetPictureList(self, uuid, action):
        picture_list = []
        cur = self.dbconn.cursor()
        try:
            if action == 'moments':
                cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                      m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                               FROM ZGENERICASSET m, ZADDITIONALASSETATTRIBUTES b
                               WHERE m.ZCLOUDDELETESTATE = 0
                                 AND m.ZCOMPLETE = 1
                                 AND m.ZHIDDEN = 0
                                 AND m.ZTRASHEDSTATE = 0
                                 AND b.ZASSET = m.Z_PK
                                 AND m.ZUUID = ?
                               GROUP BY m.ZUUID
                               ORDER BY m.ZDATECREATED ASC
                               """, (uuid,))
            elif action == 'search_by_year':
                (year) = uuid
                cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                      m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                               FROM ZGENERICASSET m, ZADDITIONALASSETATTRIBUTES b
                               WHERE m.ZCLOUDDELETESTATE = 0
                                 AND m.ZCOMPLETE = 1
                                 AND m.ZHIDDEN = 0
                                 AND m.ZTRASHEDSTATE = 0
                                 AND b.ZASSET = m.Z_PK
                                 AND strftime(
                                   '%Y', t + 978307200, 'unixepoch'
                                   ) = ?
                               GROUP BY m.ZUUID
                               ORDER BY m.ZDATECREATED ASC
                               """, ('%s' % (year),))
            elif action == 'search_by_month':
                (year, month) = uuid
                cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                      m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                               FROM ZGENERICASSET m, ZADDITIONALASSETATTRIBUTES b
                               WHERE m.ZCLOUDDELETESTATE = 0
                                 AND m.ZCOMPLETE = 1
                                 AND m.ZHIDDEN = 0
                                 AND m.ZTRASHEDSTATE = 0
                                 AND b.ZASSET = m.Z_PK
                                 AND strftime(
                                   '%Y-%m', t + 978307200, 'unixepoch'
                                   ) = ?
                               GROUP BY m.ZUUID
                               ORDER BY m.ZDATECREATED ASC
                               """, ('%s-%s' % (year, month),))
            elif action == 'search_by_day':
                (year, month, day) = uuid
                cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                      m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                               FROM ZGENERICASSET m, ZADDITIONALASSETATTRIBUTES b
                               WHERE m.ZCLOUDDELETESTATE = 0
                                 AND m.ZCOMPLETE = 1
                                 AND m.ZHIDDEN = 0
                                 AND m.ZTRASHEDSTATE = 0
                                 AND b.ZASSET = m.Z_PK
                                 AND strftime(
                                   '%Y-%m-%d', t + 978307200, 'unixepoch'
                                   ) = ?
                               GROUP BY m.ZUUID
                               ORDER BY m.ZDATECREATED ASC
                               """, ('%s-%s-%s' % (year, month, day),))
            elif action == 'search_by_timestamp':
                (timestamp) = uuid
                cur1 = self.dbconn.cursor()
                cur1.execute("""SELECT strftime('%Y-%m-%d', ? + 978307200, 'unixepoch')""",
                             ('%d' % (int(float(timestamp))),))
                (date) = cur1.fetchone()
                cur1.close()
                cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                      m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                               FROM ZGENERICASSET m, ZADDITIONALASSETATTRIBUTES b
                               WHERE m.ZCLOUDDELETESTATE = 0
                                 AND m.ZCOMPLETE = 1
                                 AND m.ZHIDDEN = 0
                                 AND m.ZTRASHEDSTATE = 0
                                 AND b.ZASSET = m.Z_PK
                                 AND strftime(
                                   '%Y-%m-%d', t + 978307200, 'unixepoch'
                                   ) = ?
                               GROUP BY m.ZUUID
                               ORDER BY m.ZDATECREATED ASC
                               """, (date[0],))
            else:
                cur.execute("""SELECT (m.ZDATECREATED + b.ZTIMEZONEOFFSET) AS t,
                                      m.ZDIRECTORY, m.ZFILENAME, m.ZHASADJUSTMENTS
                               FROM ZGENERICASSET m, Z_26ASSETS n,
                                    ZGENERICALBUM a, ZADDITIONALASSETATTRIBUTES b
                               WHERE n.Z_26ALBUMS = a.Z_PK
                                 AND m.Z_PK = n.Z_34ASSETS
                                 AND m.ZCLOUDDELETESTATE = 0
                                 AND m.ZCOMPLETE = 1
                                 AND m.ZHIDDEN = 0
                                 AND m.ZTRASHEDSTATE = 0
                                 AND b.ZASSET = m.Z_PK
                                 AND a.ZUUID = ?
                               GROUP BY m.ZUUID
                               ORDER BY m.ZDATECREATED ASC
                               """, (uuid,))
            for row in cur:
                picture_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.db: GetPictureList: ' + e, xbmc.LOGERROR)
        cur.close()
        return picture_list


resources/lib/common.py
# -*- coding: utf-8 -*-

import datetime, time
import xbmc

def log_notice(msg=''):
    xbmc.log(smart_utf8(msg), xbmc.LOGNOTICE)

def log_error(msg=''):
    xbmc.log(smart_utf8(msg), xbmc.LOGERROR)

def notify(msg=''):
    xbmc.executebuiltin('XBMC.Notification(%s,%s,3000)' % ('Photos Viewer', smart_utf8(msg)))

def smart_unicode(s):
    """credit : sfaxman"""
    if not s:
        return ''
    try:
        if not isinstance(s, basestring):
            if hasattr(s, '__unicode__'):
                s = unicode(s)
            else:
                s = unicode(str(s), 'UTF-8')
        elif not isinstance(s, unicode):
            s = unicode(s, 'UTF-8')
    except:
        if not isinstance(s, basestring):
            if hasattr(s, '__unicode__'):
                s = unicode(s)
            else:
                s = unicode(str(s), 'ISO-8859-1')
        elif not isinstance(s, unicode):
            s = unicode(s, 'ISO-8859-1')
    return s

def smart_utf8(s):
    return smart_unicode(s).encode('utf-8')


resources/lib/db_viewmode.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals
from future import standard_library
standard_library.install_aliases()
from builtins import *
from builtins import object
try:
    from sqlite3 import dbapi2 as sqlite
except Exception:
    from pysqlite2 import dbapi2 as sqlite

import xbmc

class ViewModeDB(object):

    def __init__(self, dbfile):
        self.OpenDB(dbfile)

    def __del__(self):
        self.CloseDB()

    def OpenDB(self, dbfile):
        try:
            self.dbconn = sqlite.connect(dbfile)
        except Exception:
            pass

    def CloseDB(self):
        try:
            self.dbconn.close()
        except Exception:
            pass

    def GetViewMode(self, url_current):
        mode_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute("""SELECT path FROM view WHERE path = ?""", (url_current,))
            for row in cur:
                mode_list.append(row)
        except Exception as e:
            xbmc.log('photosviewer.viewmode_db: GetViewMode: ' + e, xbmc.LOGERROR)
        cur.close()
        return mode_list


addon.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals
from future import standard_library
standard_library.install_aliases()
from builtins import *
from builtins import object
import sys
import os
import urllib.request, urllib.parse, urllib.error
import datetime
import re

import xbmc
import xbmcgui as gui
import xbmcplugin as plugin
import xbmcaddon
import xbmcvfs

from resources.lib.db import *
from resources.lib.db_viewmode import *

addon = xbmcaddon.Addon()

plugin_path = addon.getAddonInfo('path')
resource_path = os.path.join(plugin_path, 'resources')
lib_path = os.path.join(resource_path, 'lib')
sys.path.append(lib_path)

base_url = sys.argv[0]
addon_handle = int(sys.argv[1])
args = urllib.parse.parse_qs(sys.argv[2][1:])

def build_url(query):
    return base_url + '?' + urllib.parse.urlencode(query)

def convert_timestamp(year=None, month=None, day=None,
                      hour=None, minute=None, timestamp=None):
    if timestamp:
        # Use utcfromtimestamp because timestamp already has timezone offset.
        t = datetime.datetime.utcfromtimestamp(int(timestamp) + 978307200)
        (year, month, day, hour, minute) = t.replace(year=t.year).strftime('%Y,%m,%d,%H,%M').split(',')
    if year:
        if month:
            m = addon.getLocalizedString(30017).split(',')
            mindex = int(month) - 1
            if day:
                d = addon.getLocalizedString(30018).split(',')
                dindex = datetime.date(int(year), int(month), int(day)).weekday()
                if hour and minute:
                    itemname = addon.getLocalizedString(30030).format(year=year, month=m[mindex],
                                                                      day=day, day7=d[dindex],
                                                                      hour=hour, minute=minute)
                else:
                    itemname = addon.getLocalizedString(30031).format(year=year, month=m[mindex],
                                                                      day=day, day7=d[dindex])
                if dindex == 5 or dindex == 6:
                    itemname = '[COLOR blue]' + itemname + '[/COLOR]'
            else:
                itemname = addon.getLocalizedString(30032).format(year=year, month=m[mindex])
        else:
            itemname = addon.getLocalizedString(30033).format(year=year)
    else:
        itemname = ''
    return itemname

class App(object):

    def __init__(self):

        self.params = {}
        qs = urllib.parse.parse_qs(urllib.parse.urlparse(sys.argv[2]).query, keep_blank_values=True)
        for key in list(qs.keys()):
            self.params[key] = qs[key][0]

        self.photo_app_path = addon.getSetting('photo_library_path')
        if self.photo_app_path == '':
            self.photo_app_path = os.path.join(
                os.getenv('HOME'), 'Pictures', addon.getLocalizedString(30000))
        addon.setSetting('photo_library_path', self.photo_app_path)
        self.display_adjusted = addon.getSetting('display_adjusted')

        self.photo_app_db_file = os.path.join(
            xbmc.translatePath(addon.getAddonInfo('profile')), 'Photos.sqlite')
        self.photo_app_db_wal_file = os.path.join(
            xbmc.translatePath(addon.getAddonInfo('profile')), 'Photos.sqlite-wal')
        self.photo_app_db_orig = os.path.join(
            self.photo_app_path, 'database', 'Photos.sqlite')
        self.photo_app_db_wal_orig = os.path.join(
            self.photo_app_path, 'database', 'Photos.sqlite-wal')
        ctime = os.stat(self.photo_app_db_orig).st_ctime
        mtime = os.stat(self.photo_app_db_orig).st_mtime
        if xbmcvfs.exists(self.photo_app_db_wal_orig):
            ctime_wal = os.stat(self.photo_app_db_wal_orig).st_ctime
            mtime_wal = os.stat(self.photo_app_db_wal_orig).st_mtime
            if (xbmcvfs.exists(self.photo_app_db_file)
                    and os.stat(self.photo_app_db_file).st_mtime >= mtime_wal):
                pass
            else:
                xbmcvfs.copy(self.photo_app_db_orig, self.photo_app_db_file)
                os.utime(self.photo_app_db_file, (ctime, mtime))
                xbmcvfs.copy(self.photo_app_db_wal_orig, self.photo_app_db_wal_file)
                os.utime(self.photo_app_db_wal_file, (ctime_wal, mtime_wal))
        else:
            if (xbmcvfs.exists(self.photo_app_db_file)
                    and os.stat(self.photo_app_db_file).st_mtime >= mtime):
                pass
            else:
                xbmcvfs.copy(self.photo_app_db_orig, self.photo_app_db_file)
                os.utime(self.photo_app_db_file, (ctime, mtime))

        self.photo_app_picture_path = os.path.join(self.photo_app_path, 'originals')
        self.photo_app_rendered_path = os.path.join(self.photo_app_path, 'resources', 'renders')

        self.viewmode_db_file = os.path.join(xbmc.translatePath('special://database'), 'ViewModes6.db')

        self.db = None
        self.viewmode_db = None
        self.view_mode = 0

    def open_db(self):
        if self.db is not None: return
        try:
            self.db = DB(self.photo_app_db_file)
        except Exception:
            pass

    def close_db(self):
        try:
            self.db.CloseDB()
        except Exception:
            pass

    def open_viewmode_db(self):
        if self.viewmode_db is not None: return
        try:
            self.viewmode_db = ViewModeDB(self.viewmode_db_file)
        except Exception:
            pass

    def close_viewmode_db(self):
        try:
            self.viewmode_db.CloseDB()
        except Exception:
            pass

    def set_viewmode(self):
        view_mode = int(addon.getSetting('view_mode'))
        if (view_mode == 1):
            self.view_mode = 500    # Wall
        elif (view_mode == 2):
            self.view_mode = 53     # Shift
        elif (view_mode == 3):
            self.view_mode = 54     # InfoWall
        else:
            self.view_mode = 55     # WideList

    def list_moments(self, year, month, url_current):
        n = 0
        moments = self.db.GetMomentList(year, month)
        for (name,) in moments:
            if year is None:
                url = build_url({'action': 'moments', 'year': name})
                item = gui.ListItem(convert_timestamp(year=name))
                item.setArt({'icon': 'DefaultYear.png', 'thumb': 'DefaultYear.png'})
                contextmenu = []
                contextmenu.append((
                    addon.getLocalizedString(30012).format(
                        period=convert_timestamp(year=name)),
                    'XBMC.Container.Update(%s)' % build_url(
                        {'action': 'search_by_year', 'year': name})))
                item.addContextMenuItems(contextmenu)
            elif month is None:
                url = build_url({'action': 'moments', 'year': year[0], 'month': name})
                item = gui.ListItem(convert_timestamp(year=year[0], month=name))
                item.setArt({'icon': 'DefaultYear.png', 'thumb': 'DefaultYear.png'})
                contextmenu = []
                contextmenu.append((
                    addon.getLocalizedString(30012).format(
                        period=convert_timestamp(year=year[0],month=name)),
                    'XBMC.Container.Update(%s)' % build_url(
                        {'action': 'search_by_month', 'year': year[0], 'month': name})))
                contextmenu.append((
                    addon.getLocalizedString(30012).format(
                        period=convert_timestamp(year=year[0])),
                    'XBMC.Container.Update(%s)' % build_url(
                        {'action': 'search_by_year', 'year': year[0]})))
                item.addContextMenuItems(contextmenu)
            else:
                url = build_url({'action': 'search_by_day',
                                 'year': year[0], 'month': month[0], 'day': name})
                item = gui.ListItem(convert_timestamp(year=year[0],month=month[0],day=name))
                item.setArt({'icon': 'DefaultYear.png', 'thumb': 'DefaultYear.png'})
                contextmenu = []
                contextmenu.append((
                    addon.getLocalizedString(30012).format(
                        period=convert_timestamp(year=year[0],month=month[0])),
                    'XBMC.Container.Update(%s)' % build_url(
                        {'action': 'search_by_month', 'year': year[0], 'month': month[0]})))
                contextmenu.append((
                    addon.getLocalizedString(30012).format(
                        period=convert_timestamp(year=year[0])),
                    'XBMC.Container.Update(%s)' % build_url(
                        {'action': 'search_by_year', 'year': year[0]})))
                item.addContextMenuItems(contextmenu)
            plugin.addDirectoryItem(addon_handle, url, item, True)
            n += 1
        if not self.viewmode_db.GetViewMode(url_current):
            self.view_mode = 52    # IconWall
        return n

    def list_albums(self, folderUuid, url_current):
        n = 0
        folders = self.db.GetFolderList(folderUuid)
        for (name, uuid) in folders:
            url = build_url({'action': 'albums', 'folderUuid': uuid})
            item = gui.ListItem(name)
            item.setArt({'icon': 'DefaultFolder.png', 'thumb': 'DefaultFolder.png'})
            plugin.addDirectoryItem(addon_handle, url, item, True)
            n += 1
        albums = self.db.GetAlbumList(folderUuid)
        for (name, uuid) in albums:
            url = build_url({'action': 'albums', 'uuid': uuid})
            item = gui.ListItem(name)
            item.setArt({'icon': 'DefaultPicture.png', 'thumb': 'DefaultPicture.png'})
            plugin.addDirectoryItem(addon_handle, url, item, True)
            n += 1
        if not self.viewmode_db.GetViewMode(url_current):
            self.view_mode = 52    # IconWall
        return n

    def list_slideshows(self, url_current):
        n = 0
        slideshows = self.db.GetSlideshowList()
        for (name, uuid) in slideshows:
            url = build_url({'action': 'slideshows', 'uuid': uuid})
            item = gui.ListItem(name)
            item.setArt({'icon': 'DefaultPicture.png', 'thumb': 'DefaultPicture.png'})
            plugin.addDirectoryItem(addon_handle, url, item, True)
            n += 1
        if not self.viewmode_db.GetViewMode(url_current):
            self.view_mode = 52    # IconWall
        return n

    def list_photos(self, uuid, action, url_current):
        pictures = self.db.GetPictureList(uuid, action)
        n = 0
        for (imageDate, imagePath, imageFilename, isAdjusted) in pictures:
            if isAdjusted == 1 and self.display_adjusted == 'true':
                imagePath = os.path.join(
                    self.photo_app_rendered_path, imagePath,
                    re.sub(r'^([-A-Z0-9]*)\.', r'\g<1>_1_201_a.', imageFilename))
            else:
                imagePath = os.path.join(
                    self.photo_app_picture_path, imagePath, imageFilename)
            item = gui.ListItem(convert_timestamp(timestamp=imageDate))
            contextmenu = []
            contextmenu.append((
                addon.getLocalizedString(30010),
                'XBMC.Container.Update(%s)' % build_url(
                    {'action': 'search_by_timestamp', 'timestamp': imageDate})))
            contextmenu.append((
                addon.getLocalizedString(30014),
                'XBMC.Container.Update(%s)' % build_url({})))
            item.addContextMenuItems(contextmenu)
            plugin.addDirectoryItem(addon_handle, imagePath, item, False)
            n += 1
        plugin.setContent(addon_handle, 'images')
        if not self.viewmode_db.GetViewMode(url_current):
            self.set_viewmode()
        return n

    def list_videos(self, url_current):
        n = 0
        videos = self.db.GetVideoList()
        for (imageDate, imagePath, imageFilename, isAdjusted) in videos:
            if isAdjusted == 1 and self.display_adjusted == 'true':
                imagePath = os.path.join(
                    self.photo_app_rendered_path, imagePath,
                    re.sub(r'^([-A-Z0-9]*)\.', r'\g<1>_2_0_a.', imageFilename))
            else:
                imagePath = os.path.join(
                    self.photo_app_picture_path, imagePath, imageFilename)
            item = gui.ListItem(convert_timestamp(timestamp=imageDate))
            contextmenu = []
            contextmenu.append((
                addon.getLocalizedString(30010),
                'XBMC.Container.Update(%s)' % build_url(
                    {'action': 'search_by_timestamp', 'timestamp': imageDate})))
            contextmenu.append((
                addon.getLocalizedString(30014),
                'XBMC.Container.Update(%s)' % build_url({})))
            item.addContextMenuItems(contextmenu)
            plugin.addDirectoryItem(addon_handle, imagePath, item, False)
            n += 1
        plugin.setContent(addon_handle, 'videos')
        if not self.viewmode_db.GetViewMode(url_current):
            self.set_viewmode()
        return n

    def main_menu(self):
        url = build_url({'action': 'moments'})
        item = gui.ListItem(addon.getLocalizedString(30001))
        item.setArt({'icon': 'DefaultPicture.png', 'thumb': 'DefaultPicture.png'})
        plugin.addDirectoryItem(addon_handle, url, item, True)

        url = build_url({'action': 'albums', 'folderUuid': 'root'})
        item = gui.ListItem(addon.getLocalizedString(30004))
        item.setArt({'icon': 'DefaultFolder.png', 'thumb': 'DefaultFolder.png'})
        plugin.addDirectoryItem(addon_handle, url, item, True)

        url = build_url({'action': 'videos'})
        item = gui.ListItem(addon.getLocalizedString(30005))
        item.setArt({'icon': 'DefaultVideo.png', 'thumb': 'DefaultVideo.png'})
        plugin.addDirectoryItem(addon_handle, url, item, True)

        url = build_url({'action': 'slideshows'})
        item = gui.ListItem(addon.getLocalizedString(30006))
        item.setArt({'icon': 'DefaultFolder.png', 'thumb': 'DefaultFolder.png'})
        plugin.addDirectoryItem(addon_handle, url, item, True)

        self.view_mode = 55    # WideList

        return 4

if __name__ == '__main__':

    action_result = None
    items = 0

    url_current = sys.argv[0] + sys.argv[2]
    action = args.get('action', None)
    folderUuid = args.get('folderUuid', None)
    uuid = args.get('uuid', None)
    year = args.get('year', None)
    month = args.get('month', None)
    day = args.get('day', None)
    timestamp = args.get('timestamp', None)

    app = App()
    app.open_db()
    app.open_viewmode_db()

    if action is None:
        items = app.main_menu()
    elif not (uuid is None):
        items = app.list_photos(uuid[0], action[0], url_current)
    elif action[0] == 'moments':
        items = app.list_moments(year, month, url_current)
    elif action[0] == 'albums':
        items = app.list_albums(folderUuid[0], url_current)
    elif action[0] == 'videos':
        items = app.list_videos(url_current)
    elif action[0] == 'slideshows':
        items = app.list_slideshows(url_current)

    elif action[0] == 'search_by_year':
        items = app.list_photos((year[0]), action[0], url_current)
    elif action[0] == 'search_by_month':
        items = app.list_photos((year[0], month[0]), action[0], url_current)
    elif action[0] == 'search_by_day':
        items = app.list_photos((year[0], month[0], day[0]), action[0], url_current)
    elif action[0] == 'search_by_timestamp':
        items = app.list_photos((timestamp[0]), action[0], url_current)

    app.close_db()
    app.close_viewmode_db()

    if items == 0:
        action_result = addon.getLocalizedString(30100)
    else:
        plugin.endOfDirectory(addon_handle, True)
        xbmc.sleep(300)
        if app.view_mode:
            xbmc.executebuiltin('Container.SetViewMode(%d)' % app.view_mode)
    
    if action_result:
        xbmc.executebuiltin(b'XBMC.Notification(%s,%s,3000)' % (b'Photos Viewer', action_result.encode('utf-8')))
